File: runtime/java/prompty/src/main/java/com/microsoft/prompty/Memory.java
package com.microsoft.prompty;

import com.microsoft.prompty.model.MemoryCategory;
import com.microsoft.prompty.model.MemoryEntry;
import com.microsoft.prompty.model.MemoryStore;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

/**
 * Agent memory: tiered recall, formatting, and mutation over the generated memory contract.
 *
 * <p>The data types are generated; what a runtime must not reinvent is what the data <em>means</em>.
 * That is what lives here — ranking, injection, eviction — so a host is left owning only
 * persistence, and every runtime agrees on which memory a query should surface.
 *
 * <p>Memory is tiered. A {@link MemoryCategory#CORE} memory is a persistent fact: injected into
 * every system prompt, replaced on write when an identical tag set restates it, and boosted during
 * recall. A {@link MemoryCategory#ARCHIVAL} memory is a compressed summary, surfaced only through
 * recall and the first thing evicted when the store is full. A {@link MemoryCategory#INSIGHT}
 * memory is a saved reflection, surfaced through recall.
 *
 * <p>Associating a memory with a session, a user, or a project is host convention expressed through
 * the general {@code tags} field, not engine logic.
 *
 * <p>These operations mutate the store in place, because Java cannot add methods to the generated
 * type the way the reference runtime can.
 */
public final class Memory {

  private Memory() {}

  /** A recalled memory paired with the score that ranked it. */
  public record Scored(MemoryEntry entry, double score, int keywordMatches) {}

  /**
   * Host-owned persistence for a whole-store snapshot.
   *
   * <p>Deliberately whole-store rather than per-entry: a host that persists entries individually
   * has to reason about ordering, which is exactly the deterministic logic this module owns.
   */
  public interface Port {
    MemoryStore load();

    void save(MemoryStore store);
  }

  /** Append a memory with no tier policy applied. */
  public static void add(MemoryStore store, MemoryEntry entry) {
    store.entries.add(entry);
  }

  /**
   * Insert a memory applying tier policy.
   *
   * <p>A restated core fact replaces the old one rather than accumulating beside it, which is what
   * keeps an always-injected tier from growing without bound. Identity for that purpose is the tag
   * set, since that is what a host uses to scope a fact.
   *
   * <p>A {@code maxEntries} of {@code 0} means no cap.
   */
  public static void remember(MemoryStore store, MemoryEntry entry, int maxEntries) {
    if (entry.category == MemoryCategory.CORE) {
      store.entries.removeIf(
          existing -> existing.category == MemoryCategory.CORE && tagsEqual(existing.tags, entry.tags));
    }
    store.entries.add(entry);
    evictToCap(store, maxEntries);
  }

  /**
   * Evict until the store holds at most {@code maxEntries}, and report how many went.
   *
   * <p>Archival memories go first because they are summaries of things already said; only when
   * there are none does the oldest memory of any tier go. A {@code maxEntries} of {@code 0} means
   * no cap and evicts nothing.
   */
  public static int evictToCap(MemoryStore store, int maxEntries) {
    if (maxEntries <= 0) {
      return 0;
    }
    int evicted = 0;
    while (store.entries.size() > maxEntries) {
      int victim = 0;
      for (int i = 0; i < store.entries.size(); i++) {
        if (store.entries.get(i).category == MemoryCategory.ARCHIVAL) {
          victim = i;
          break;
        }
      }
      store.entries.remove(victim);
      evicted++;
    }
    return evicted;
  }

  /** Replace the memory at {@code index}. */
  public static void update(MemoryStore store, int index, MemoryEntry entry) {
    requireIndex(store, index);
    store.entries.set(index, entry);
  }

  /** Replace only the content at {@code index}, preserving category, timestamp, and tags. */
  public static void updateContent(MemoryStore store, int index, String content) {
    requireIndex(store, index);
    store.entries.get(index).content = content;
  }

  /** Remove and return the memory at {@code index}. */
  public static MemoryEntry remove(MemoryStore store, int index) {
    requireIndex(store, index);
    return store.entries.remove(index);
  }

  /** Remove memories in {@code category}, or every memory when it is null, and report how many. */
  public static int clear(MemoryStore store, MemoryCategory category) {
    int before = store.entries.size();
    if (category == null) {
      store.entries.clear();
    } else {
      store.entries.removeIf(entry -> entry.category == category);
    }
    return before - store.entries.size();
  }

  /** The core memories, in insertion order. */
  public static List<MemoryEntry> coreMemories(MemoryStore store) {
    List<MemoryEntry> core = new ArrayList<>();
    for (MemoryEntry entry : store.entries) {
      if (entry.category == MemoryCategory.CORE) {
        core.add(entry);
      }
    }
    return core;
  }

  /**
   * Recall the most relevant memories for {@code query}.
   *
   * <p>Ranking is lexical and dependency-free, so it produces the same order in every runtime and
   * on every machine. A keyword found in the content is worth 2; found in the tags it is worth 3,
   * because tags were chosen deliberately and content merely happens to contain the word. A core
   * memory that matched at all gets a further 1, being the always-relevant tier. Ties fall back to
   * insertion order.
   *
   * <p>A query with no keywords returns everything in insertion order at score 0; a {@code limit}
   * of {@code 0} returns every match. A host wanting embedding-based recall keeps the vectors in
   * its own storage and does that itself.
   */
  public static List<Scored> recall(MemoryStore store, String query, int limit) {
    List<String> tokens = queryTokens(query);
    boolean hasQuery = !tokens.isEmpty();

    List<Scored> scored = new ArrayList<>();
    for (MemoryEntry entry : store.entries) {
      double[] result = scoreEntry(entry, tokens);
      int matches = (int) result[1];
      if (hasQuery && matches == 0) {
        continue;
      }
      scored.add(new Scored(entry, result[0], matches));
    }

    // List.sort is contractually stable, which is what keeps equal scores in insertion order.
    scored.sort(Comparator.comparingDouble(Scored::score).reversed());

    return limit > 0 && scored.size() > limit ? List.copyOf(scored.subList(0, limit)) : scored;
  }

  /**
   * Format the core memories as a block for injection into a system prompt.
   *
   * <p>Returns an empty string when there are none, so a host can inject conditionally without
   * first asking. Only core memories are injected; the other tiers reach the model through recall.
   */
  public static String formatForSystemPrompt(MemoryStore store) {
    List<MemoryEntry> core = coreMemories(store);
    if (core.isEmpty()) {
      return "";
    }
    StringBuilder out = new StringBuilder("## Memory\n");
    for (MemoryEntry entry : core) {
      out.append("- ").append(entry.content).append('\n');
    }
    return out.toString();
  }

  /**
   * Format a recall result set for presentation, e.g. to show which memories informed a response.
   *
   * <p>Returns an empty string for an empty result set.
   */
  public static String formatRecallResults(List<Scored> results) {
    if (results.isEmpty()) {
      return "";
    }
    StringBuilder out = new StringBuilder();
    for (int i = 0; i < results.size(); i++) {
      Scored scored = results.get(i);
      out.append(i + 1)
          .append(". [")
          .append(scored.entry().category.value)
          .append("] ")
          .append(scored.entry().content)
          .append('\n');
      List<String> tags = scored.entry().tags;
      if (tags != null && !tags.isEmpty()) {
        out.append("   tags: ").append(String.join(", ", tags)).append('\n');
      }
    }
    return out.toString();
  }

  private static void requireIndex(MemoryStore store, int index) {
    if (index < 0 || index >= store.entries.size()) {
      throw new IndexOutOfBoundsException(
          "memory index " + index + " out of bounds (len " + store.entries.size() + ")");
    }
  }

  /** Whether two tag lists carry the same tags, treating an absent list and an empty one as equal. */
  private static boolean tagsEqual(List<String> left, List<String> right) {
    List<String> a = left == null ? List.of() : left;
    List<String> b = right == null ? List.of() : right;
    return Objects.equals(a, b);
  }

  private static List<String> queryTokens(String query) {
    List<String> tokens = new ArrayList<>();
    if (query == null) {
      return tokens;
    }
    for (String raw : query.split("\\s+")) {
      String token = trimNonAlphanumeric(raw).toLowerCase(Locale.ROOT);
      if (!token.isEmpty() && !tokens.contains(token)) {
        tokens.add(token);
      }
    }
    return tokens;
  }

  private static String trimNonAlphanumeric(String value) {
    int start = 0;
    int end = value.length();
    while (start < end && !Character.isLetterOrDigit(value.charAt(start))) {
      start++;
    }
    while (end > start && !Character.isLetterOrDigit(value.charAt(end - 1))) {
      end--;
    }
    return value.substring(start, end);
  }

  /** Returns {@code [weightedScore, distinctMatches]}; both are zero for an empty query. */
  private static double[] scoreEntry(MemoryEntry entry, List<String> tokens) {
    if (tokens.isEmpty()) {
      return new double[] {0.0, 0};
    }
    String content = entry.content == null ? "" : entry.content.toLowerCase(Locale.ROOT);
    List<String> tags = new ArrayList<>();
    if (entry.tags != null) {
      for (String tag : entry.tags) {
        tags.add(tag == null ? "" : tag.toLowerCase(Locale.ROOT));
      }
    }

    double weighted = 0.0;
    int distinct = 0;
    for (String token : tokens) {
      boolean inContent = content.contains(token);
      boolean inTags = tags.stream().anyMatch(tag -> tag.contains(token));
      if (inContent || inTags) {
        distinct++;
      }
      if (inContent) {
        weighted += 2.0;
      }
      if (inTags) {
        weighted += 3.0;
      }
    }
    if (weighted > 0.0 && entry.category == MemoryCategory.CORE) {
      weighted += 1.0;
    }
    return new double[] {weighted, distinct};
  }
}

File: runtime/python/prompty/prompty/memory.py
"""Provide host-neutral agent memory behavior.

This module layers deterministic recall, formatting, mutation, and snapshot
persistence helpers on top of the generated memory data contract. Hosts own only
where a memory snapshot is stored; the runtime owns the portable behavior.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Protocol

from .model import MemoryEntry, MemoryStore

__all__ = [
    "MemoryEntry",
    "MemoryPort",
    "MemoryStore",
    "ScoredMemory",
    "add_memory",
    "clear_memory",
    "core_memories",
    "evict_to_cap",
    "format_for_system_prompt",
    "format_recall_results",
    "recall",
    "remember",
    "remove_memory",
    "update_content",
    "update_memory",
]

CORE = "core"
ARCHIVAL = "archival"


class MemoryPort(Protocol):
    """Host-owned persistence for a whole-store memory snapshot."""

    def load(self) -> MemoryStore:
        """Load the whole memory store snapshot."""

    def save(self, store: MemoryStore) -> None:
        """Persist the whole memory store snapshot."""


@dataclass(frozen=True)
class ScoredMemory:
    """A recalled memory paired with its deterministic relevance score."""

    entry: MemoryEntry
    score: float
    keyword_matches: int


def _query_tokens(query: str) -> list[str]:
    tokens: list[str] = []
    for raw in query.split():
        token = re.sub(r"^\W+|\W+$", "", raw, flags=re.UNICODE).lower()
        if token and token not in tokens:
            tokens.append(token)
    return tokens


def _tags_equal(left: list[str] | None, right: list[str] | None) -> bool:
    return (left or []) == (right or [])


def _score_entry(entry: MemoryEntry, tokens: list[str]) -> tuple[float, int]:
    if not tokens:
        return 0.0, 0

    content = entry.content.lower()
    tags = [tag.lower() for tag in (entry.tags or [])]

    weighted = 0.0
    distinct = 0
    for token in tokens:
        in_content = token in content
        in_tags = any(token in tag for tag in tags)
        if in_content or in_tags:
            distinct += 1
        if in_content:
            weighted += 2.0
        if in_tags:
            weighted += 3.0

    if weighted > 0.0 and entry.category == CORE:
        weighted += 1.0
    return weighted, distinct


def add_memory(store: MemoryStore, entry: MemoryEntry) -> None:
    """Append a memory to the end of the store with no tier policy applied."""

    store.entries.append(entry)


def remember(store: MemoryStore, entry: MemoryEntry, max_entries: int = 0) -> None:
    """Insert a memory applying core deduplication and cap eviction."""

    if entry.category == CORE:
        store.entries = [
            existing
            for existing in store.entries
            if existing.category != CORE or not _tags_equal(existing.tags, entry.tags)
        ]
    store.entries.append(entry)
    evict_to_cap(store, max_entries)


def evict_to_cap(store: MemoryStore, max_entries: int) -> int:
    """Evict to ``max_entries``, preferring oldest archival memories."""

    if max_entries == 0:
        return 0

    evicted = 0
    while len(store.entries) > max_entries:
        victim = next((i for i, entry in enumerate(store.entries) if entry.category == ARCHIVAL), 0)
        del store.entries[victim]
        evicted += 1
    return evicted


def update_memory(store: MemoryStore, index: int, entry: MemoryEntry) -> None:
    """Replace the memory at ``index``."""

    if index >= len(store.entries):
        raise IndexError(f"memory index {index} out of bounds (len {len(store.entries)})")
    store.entries[index] = entry


def update_content(store: MemoryStore, index: int, content: str) -> None:
    """Replace only the content of the memory at ``index``."""

    if index >= len(store.entries):
        raise IndexError(f"memory index {index} out of bounds (len {len(store.entries)})")
    store.entries[index].content = content


def remove_memory(store: MemoryStore, index: int) -> MemoryEntry:
    """Remove and return the memory at ``index``."""

    if index >= len(store.entries):
        raise IndexError(f"memory index {index} out of bounds (len {len(store.entries)})")
    return store.entries.pop(index)


def clear_memory(store: MemoryStore, category: str | None = None) -> int:
    """Remove memories in ``category`` or all memories when omitted."""

    before = len(store.entries)
    if category is None:
        store.entries.clear()
    else:
        store.entries = [entry for entry in store.entries if entry.category != category]
    return before - len(store.entries)


def core_memories(store: MemoryStore) -> list[MemoryEntry]:
    """Return core-tier memories in insertion order."""

    return [entry for entry in store.entries if entry.category == CORE]


def recall(store: MemoryStore, query: str, limit: int = 0) -> list[ScoredMemory]:
    """Deterministically recall the most relevant memories for ``query``."""

    tokens = _query_tokens(query)
    has_query = bool(tokens)
    scored: list[tuple[int, ScoredMemory]] = []

    for index, entry in enumerate(store.entries):
        score, matches = _score_entry(entry, tokens)
        if has_query and matches == 0:
            continue
        scored.append((index, ScoredMemory(entry=entry, score=score, keyword_matches=matches)))

    scored.sort(key=lambda item: (-item[1].score, item[0]))
    results = [item[1] for item in scored]
    if limit > 0:
        return results[:limit]
    return results


def format_for_system_prompt(store: MemoryStore) -> str:
    """Format core memories as a deterministic system-prompt block."""

    core = core_memories(store)
    if not core:
        return ""
    return "## Memory\n" + "".join(f"- {entry.content}\n" for entry in core)


def format_recall_results(results: list[ScoredMemory]) -> str:
    """Format recalled memories for presentation."""

    if not results:
        return ""

    lines: list[str] = []
    for index, scored in enumerate(results, start=1):
        lines.append(f"{index}. [{scored.entry.category}] {scored.entry.content}")
        if scored.entry.tags:
            lines.append(f"   tags: {', '.join(scored.entry.tags)}")
    return "\n".join(lines) + "\n"
